lib/stdlib/src/gen_server.erl

```diff
@@ -2342,7 +2346,7 @@ try_handle_continue(#callback_cache{module = Mod}, Msg, State) ->
             {'EXIT', Class, R, Stacktrace}
     end.
 
-try_handle_info(#callback_cache{module = Mod, handle_info = HandleInfo}, Msg, State) ->
+try_handle_info(#server_data{module=Mod, callback_cache=#callback_cache{handle_info = HandleInfo}}, State, Msg) ->
     try
         {ok, HandleInfo(Msg, State)}
```

Comment:
I'm not really convinced this is an improvement.

While it does reduce traffic on the stack, this function now has to go through quadruple pointer indirection: `Arg -> server_data -> callback_cache -> Fun`. Stack traffic is pretty much always in CPU cache and should be very efficient, dereferencing pointers is rather expensive in modern computers. I think this should be thoroughly benchmarked.

Reply:
I don't insist on this particular form. The callbacks could just as well be placed directly in the `#server_data` record instead of a sub-record. With that, it should be equivalent to the current implementation.

Reply:
I guess @michalmuskala is right.  Having them as fields in the `#server_data` record is probably better.

This would increase the hibernation size by 4 words, which I think will be ok.

Note that in `wake_hib` it is not needed to update the callback cache since the cache contains export entries, and the `#server_data` record fields are kept over hibernation.

We might need to add code that asserts the type of the callback cache fields after `create_callback_cache`, or inline that function, so the type information gets propagated into the loop for efficient code generation.  The assembler output will tell.